src/parquet/column/writer.h

```diff
@@ -174,12 +183,12 @@ inline void TypedColumnWriter<DType>::WriteBatch(int64_t num_values,
   num_buffered_encoded_values_ += values_to_write;
 
   // TODO(PARQUET-591): Instead of rows as a boundary, do a size check
-  if (num_buffered_values_ >= PAGE_VALUE_COUNT) { WriteNewPage(); }
+  if ((num_buffered_values_ >= PAGE_VALUE_COUNT) && !has_dictionary_) { WriteNewPage(); }
```

Comment:
I believe other Parquet implementations accumulate all of the encoded data pages in memory prior to writing them all (possibly prefixed with a dictionary page) -- (see the "column page write store" concept in parquet-mr or the implementation of `BaseColumnWriter::Flush()` in Impala https://github.com/apache/incubator-impala/blob/master/be/src/exec/hdfs-parquet-table-writer.cc#L444).

We may scope this work in additional JIRAs but we'll need to do something similar, and also setting some kind of an optional dictionary size threshold so that when the dictionary grows to a certain size that we can switch over to plain encoding. You can imagine a worst-case scenario where you select dictionary encoding for strings in a large table and all of the observed strings are unique (so that the dictionary page could grow arbitrarily large)